[JENKINS-60410] Add stack trace suppression into core as a standard behavior.

jeffret-b opened the pull request:
See [JENKINS-60410](https://issues.jenkins-ci.org/browse/JENKINS-60410).

I propose to move the capability to suppress stack traces for non-administrator users into core and improve the behavior. Security has become a greater concern for many people in the past years, including for Jenkins administrators. Suppressing stack traces is industry standard behavior and we should align with that.

See stapler/stapler#176 for related, dependent changes.
See also  jenkinsci/jenkins-test-harness#201 for dependent changes to get the tests to pass.

### Proposed changelog entries

* Suppress error stack traces for non-administrator users as core capability. ([issue 60410](https://issues.jenkins-ci.org/browse/JENKINS-60410))

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

jeffret-b committed:
Add stack trace suppression into core as a standard.

jeffret-b committed:
Add stack trace suppression into core as a standard.

jeffret-b committed:
Clean up suppression filter.

jeffret-b committed:
Clean up suppression filter.

jvz commented on core/src/main/java/jenkins/security/SuppressionFilter.java:
```diff
@@ -0,0 +1,137 @@
+package jenkins.security;
```
Copyright header?

jvz reviewed (approved):
Code changes look reasonable. Nit: since these are new files, you can apply some automated code formatting and import optimization.

jeffret-b committed:
Minor review comments.

jeffret-b committed:
Minor review comments.

oleg-nenashev reviewed (changes requested):
Thanks for the pull request! I am -1 w.r.t such change if there is no opt-out. On my pet project instances I just have no user with `ADMINISTER` permission by default, because the instances are managed as-code. Also, it may hammer advanced users on instances and JIRA reporting to the Jenkins project. 

IMHO it needs more reviews and feedback

jeffret-b commented:
@oleg-nenashev , would you think a system property or a setting on the Configure Global Security page would be suitable here?

jtnord commented on core/src/main/java/jenkins/security/SuppressionFilter.java:
```diff
@@ -0,0 +1,160 @@
+
+    private void respondWithSimpleErrorPage(HttpServletResponse response, String errorId) throws ServletException {
+        try {
+            response.setStatus(SC_INTERNAL_SERVER_ERROR);
```
check `isCommitted()` first.  if the response is committed this will all just blow up and spam the log...
I expect many times that the response would be committed and thus this would be always doomed to fail.

jtnord commented on core/src/main/java/jenkins/security/SuppressionFilter.java:
```diff
@@ -0,0 +1,160 @@
+                }
+                throwServletException(e);
+            } else if (request instanceof HttpServletRequest && response instanceof HttpServletResponse) {
+                String errorId = logException((HttpServletRequest) request, e);
```
will cause a lot of log spam for things like client disconnections during write (which are to be considered normal).

jeffret-b replied:
Possibly. I can't recall seeing any stack traces in situations like that in the past. It could be that they don't show because of the type of disconnection. I'm not sure how to reproduce these types of problems in a meaningful way.

We could not log the exceptions. Just swallow them and present the error page. Or log them in at a different level. Or if we can detect these types of failures we could handle them differently.

If we could get some idea of what exceptions might fall into this category and how to examine them, we can see if there's something special we need to do here.

jeffret-b replied:
Maybe `Stapler.isSocketException(e)`, currently on line 87, covers that case.

jtnord commented on core/src/main/java/jenkins/security/SuppressionFilter.java:
```diff
@@ -0,0 +1,160 @@
+                return true;
+            }
+            currentException = currentException.getCause();
+        } while (currentException != null);
```
do not do this - you can have circular references as a you can set the cause of an exception to itself. (or something higher up the chain)  (for example see Throwable.printStackTrace() code)

Wadeck reviewed (changes requested):
Reproduction steps:
- `docker run --rm -it -p 8081:8080 -e ID=4389 jenkins/core-pr-tester`
- Install Matrix-auth
- Configure a user with just read access
- Log as that user
- Browse to <jenkins-url>/configureSecurity
- You got an error page

In this case the Exception is thrown by Jelly during the parsing of the page as there is an `AccessDeniedException` thrown (due to a missing permission) but then wrapped inside a ServletException. In the plugin it's only a check on the AccessDeniedException, but in your code you test if the exception contains the ACE, leading to still displaying some traces.

Second type of problem: http://[jenkins-url]/404, I got the full stack trace in a different format.

IOW we need to have tests covering the cases as it seems we do not cover them (or I miss something)

oleg-nenashev commented:
> would you think a system property or a setting on the Configure Global Security page would be suitable here?

Until there is a standard engine to manage feature flags, a UI control is a preference for me

oleg-nenashev reviewed (dismissed):
Requesting changes to remove it from my review dashboard for now

jeffret-b commented:
I like the idea of the UI control. That makes sense. I'll discuss this more with Wadeck and others and try to get back to this next week.

jeffret-b committed:
Handle two review comments.

Check isCommitted().
Guard against circular causes when checking for AccessDeniedException.

jeffret-b committed:
Handle two review comments.

Check isCommitted().
Guard against circular causes when checking for AccessDeniedException.

jeffret-b committed:
Better handle AccessDeniedException.

jeffret-b committed:
Better handle AccessDeniedException.

jeffret-b commented:
> In this case the Exception is thrown by Jelly during the parsing of the page as there is an `AccessDeniedException` thrown (due to a missing permission) but then wrapped inside a ServletException. In the plugin it's only a check on the AccessDeniedException, but in your code you test if the exception contains the ACE, leading to still displaying some traces.
Good catch! I've got that working.

> Second type of problem: http://[jenkins-url]/404, I got the full stack trace in a different format.
There isn't any requirement to have all errors presented in the same format. The requirement is to not show the stack trace. I don't see any problems with the way that page looks.
> IOW we need to have tests covering the cases as it seems we do not cover them (or I miss something)
Yes, I need to see if it's possible to put together some tests.

daniel-beck commented:
> Until there is a standard engine to manage feature flags, a UI control is a preference for me

I strongly disagree with this requirement, as it is a workaround for a very unusual setup but will add UI complexity for all users.

We also should not accept that experiencing unhandled errors currently resulting in stack traces to be part of the normal user experience; the more painful we make it to experience them, the more motivated we are to actually properly handle and prevent them.

To that end, I would even prefer to not show stack traces even to administrators, but only add a link to the Jenkins log on the error page to make it easy to figure out what went wrong.

jeffret-b commented:
I agree with Daniel. It would be much better to just make stack suppression a standard behavior in Jenkins. While showing the stack can make it easier for some Jenkins developers, it really isn't good user experience. These days especially it kind of suggests a level of product immaturity.

Showing the trace to administrators is a compromise, but still not good UX for real admins. It's more a developer convenience for developers who typically run as admin.

The lesser compromise Daniel suggests of showing the innermost exception message could make sense if we restrict that additional content to admins. A lot of content can get mixed into that message that we don't want to regularly share. I'm not really sure there's enough benefit in these messages for the compromise. It's probably better to just promote the behavior of looking up things in the log.

daniel-beck commented:
> Showing the trace to administrators is a compromise, but still not good UX for real admins. It's more a developer convenience for developers who typically run as admin.

This is a good point I thought about after posting my previous comment as well: It's not really for _administrators_ either. It's for _developers_; specifically _of Jenkins_. Sure, we're all developers of Jenkins, as well as administrators, and happy if we can save a click or two before typing out a bug report for 10 minutes or sinking half an hour into debugging the error cause; but OTOH I don't really want to see this internal mess when I'm just a _user_. Give me the ability to file a useful bug report (if I can even be bothered to), but otherwise please don't expose this internal mess.

(I remember a vaguely similar argument around the masking of forms while they're loading from years ago, with the argument being that, as developers, it's useful to see at what UI element the Jelly view stopped rendering. Yet we make _everyone_ look at the progressively built up form UI. Ugh.)

A system property could be supported, and set automatically when running `hpi:run` and `jetty-dev:run` to enable error stack traces for error messages, and that could be used in (I'm guessing fairly rare) configurations like the one Oleg mentioned too. But the default should be to hide it, and only provide admins a reference to the log so they can get there easily enough, with no new UI option clutter.

jeffret-b commented:
> A system property could be supported, and set automatically when running `hpi:run` and `jetty-dev:run` to enable error stack traces for error messages, and that could be used in (I'm guessing fairly rare) configurations like the one Oleg mentioned too. But the default should be to hide it, and only provide admins a reference to the log so they can get there easily enough, with no new UI option clutter.

This is a reasonable approach. Retains some ease for Jenkins devs, but cleans things up for regular users. Hiding the trace for regular admins makes sense, because it's not necessarily something they need to resolve.

jeffret-b committed:
First pass at adding tests.

jeffret-b committed:
First pass at adding tests.

jeffret-b committed:
Convert to not show the trace to admins.

Show it to everybody if the system property is set.

jeffret-b committed:
Convert to not show the trace to admins.

Show it to everybody if the system property is set.

jeffret-b committed:
Remove some unrelated tests.

jeffret-b committed:
Remove some unrelated tests.

jeffret-b committed:
Add test for basic exception.

And clean up a few tests.

jeffret-b committed:
Add test for basic exception.

And clean up a few tests.

jeffret-b committed:
Shift to using HttpResponse.SHOW_STACK_TRACE.

jeffret-b committed:
Shift to using HttpResponse.SHOW_STACK_TRACE.

jeffret-b committed:
Adjust comment.

jeffret-b committed:
Adjust comment.

jeffret-b committed:
Rename class for better clarity.

jeffret-b committed:
Rename class for better clarity.

jeffret-b committed:
Need a new Stapler release or incremental for these tests.

They really belong in Stapler anyway, but they're easier here.

jeffret-b committed:
Need a new Stapler release or incremental for these tests.

They really belong in Stapler anyway, but they're easier here.

jeffret-b committed:
Finish cleanup of Stapler separation.

jeffret-b committed:
Finish cleanup of Stapler separation.

jeffret-b committed:
Add a couple more tests and some comments about the error source.

jeffret-b committed:
Add a couple more tests and some comments about the error source.

daniel-beck commented on test/src/test/java/hudson/model/SimpleJobTest.java:
```diff
@@ -19,20 +19,20 @@
     @Rule
     public JenkinsRule rule = new JenkinsRule();
-    
+
```
This file contains no changes other than whitespace. Please revert, and configure your IDE to leave lines alone which you aren't editing.

jeffret-b replied:
No idea how that one got in there. I didn't intend to make any changes and my IDE isn't set to automatically make those changes.

daniel-beck commented:
I now wonder why we're even doing this using a filter. IIUC, it's what the plugin had to do because there's no pluggable error page in core.

Wouldn't it be much easier to use the oops page, but just not show an exception stack trace? I sketched this up in https://github.com/daniel-beck/jenkins/tree/suppress-stack-trace and it seems to work OK. Am I missing some exceptions that aren't going through this code path perhaps? (But I would guess those are done from Stapler, and then not handled by either the filter, or the oops page).

jeffret-b commented:
That's a cleaner approach. I prefer it. It seems to cover the test cases fine.

A few observations:
I prefer to switch based upon an system property rather than ADMINISTER permissions. Administrators are generally users and don't need / want the full stack trace on the page.

Printing the error message is potentially (somewhat) dangerous. It may contain information that it's really better not to share.

The oops page is kind of weird in several ways. It shows a "login" button, which is incorrect. When it gets inserted inside a page things can be really weird. It can break some other layout. It's possible this can give some contextual information that it shouldn't. There are some other oddities. It really ought to be its own page, as the plugin does, though it can certainly be prettied up if that's worthwhile.

I can take some of what you suggested there and re-work it here if you'd like.

jeffret-b commented:
The Access Denied page looks better and is better behaved.

daniel-beck commented:
Feedback from @Wadeck was that this is missing some exception stack traces that would be caught by the filter.

daniel-beck commented:
Ultimately the idea was to get a prettier error page. Perhaps we need to patch lib/layout/layout to be more minimal but not a "bare" page?

Or reuse some layout/design from the new login perhaps?

jeffret-b commented:
I'll look into suggestions for improved UX on this.

There are at least four different error handling pages:
1. Access Denied. This looks good and doesn't need any changes. One way to encounter this error is to log in as a user with Administer permission and try to access http://[JENKINS_HOST]/manage. See a proposed test [here](https://github.com/jenkinsci/jenkins/pull/4389/files#diff-26c4d8f69519b93250a055f73f8b0283R56). I'm not aware of any way to enable full stack trace.
2. Nonexistent page. Wadeck gives an example above, http://[jenkins-url]/404 . This is handled at the Stapler level and presents a simple, textual page. Changing this is outside the scope of this PR. I added a test for this case [here](https://github.com/jenkinsci/jenkins/pull/4389/commits/4a013111754fb5ac6113168d7538e9f854af1523#diff-b96bb444738637d89fdfb15df889a753R79), but later decided it was outside the scope of this PR. Full stack trace is enabled by setting `Dispatcher.TRACE = true`, which may be the default in some dev / testing environments.
3. Nonexistent adjunct. This is handled at the Stapler level. Currently it shows the full stack trace in a simple, textual format. I posted a companion PR at stapler/stapler#176 that would disable the trace by default. In the current proposal it could be enabled by a system property. I still have a test in this core PR [here](https://github.com/jenkinsci/jenkins/pull/4389/commits/4a013111754fb5ac6113168d7538e9f854af1523#diff-b96bb444738637d89fdfb15df889a753R118).
4. Uncaught error in Jenkins. This is the page connected with this PR. We have an ongoing discussion on what this should look like. This is typically handled by a programming error, which should be fixed. I have a test for this case [here](https://github.com/jenkinsci/jenkins/pull/4389/files#diff-26c4d8f69519b93250a055f73f8b0283R103). The current proposal is to enable this with a different system property.

jeffret-b commented:
@oleg-nenashev do you have any fundamental objections to proposal Daniel, Wadeck, and I are working through here?

Wadeck reviewed (approved):
> Dispatcher.TRACE = true

FTR it's also managed by system property: `stapler.trace`, so -Dstapler.trace=false is good as well.

> I posted a companion PR

Tested locally and it works as expected!

So all the cases I am aware of are covered ✔️

jeffret-b committed:
Add logging ID to uncaught exception handler.

jeffret-b committed:
Add logging ID to uncaught exception handler.

jeffret-b committed:
Revert inadvertent changes to SimpleJobTest.

jeffret-b committed:
Revert inadvertent changes to SimpleJobTest.

jeffret-b committed:
Stop adding filter for now.

jeffret-b committed:
Stop adding filter for now.

jeffret-b committed:
Small adjustments to oops.jelly.

Show stack trace only if system property is set.

jeffret-b committed:
Small adjustments to oops.jelly.

Show stack trace only if system property is set.

jeffret-b committed:
Fix tests.

jeffret-b committed:
Fix tests.

jeffret-b committed:
Catch embedded security exceptions.

jeffret-b committed:
Catch embedded security exceptions.

jeffret-b committed:
Remove messages that aren't used after all.

jeffret-b committed:
Remove messages that aren't used after all.

jeffret-b commented:
I adapted it to reuse the oops page and the uncaught exception handler instead of a filter as Daniel suggested. Adapted the page to only show the stack trace with a system property enabled. Clarified a couple of permission test cases.

jeffret-b committed:
Remove the unused messages.

jeffret-b committed:
Remove the unused messages.

Wadeck commented:
<details>

<summary>Screenshots</summary>

#### Without trace
![Screenshot_2020-01-22_102130_001](https://user-images.githubusercontent.com/2662497/72881839-d241a300-3d01-11ea-93b5-a59d3986b2dc.png)


#### With trace
![Screenshot_2020-01-22_102603_001](https://user-images.githubusercontent.com/2662497/72881846-d40b6680-3d01-11ea-83cf-ebe1aa0a0db2.png)


</details>

👍

timja commented on test/src/test/java/jenkins/security/StackTraceSuppressionTest.java:
```diff
@@ -0,0 +1,202 @@
+        assertThat(content, not(containsString("Caused by")));
+    }
+
+//    @Test
```
do you plan to remove these commented out tests?

jeffret-b replied:
I'm thinking of enabling them again once incrementals publishing is fixed and I can get a Stapler incremental. They really belong in Stapler but they're a change and they're easier to test here.

timja replied:
incremental publishing was fixed yesterday

jeffret-b replied:
It wasn't when I tried yesterday but it is now. I've updated to use the Stapler incremental.

timja commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -5259,6 +5266,10 @@ private static void computeVersion(ServletContext context) {
 
+
+    @Restricted(Beta.class)
+    public static boolean SHOW_STACK_TRACE = Boolean.getBoolean(Jenkins.class.getName() + ".SHOW_STACK_TRACE");
```
should there be some config that automatically sets this with `jetty:run`?

Wadeck replied:
That could be great yes (separate PR anyway?)

jeffret-b replied:
I think we should. We need to identify the places it should go. I expect they would all be different PRs.

jtnord replied:
why does this need to be public?  its controlled either via a system property at startup (no need to be public for that) - or you at run time via goovy no doubt (which again does not need to be public).

jeffret-b replied:
Makes testing easier. There are many other places in Jenkins that use the same pattern.

jtnord replied:
many places causing huge issues.  do not do it untill you have played with the bytecode comparability transformer

jeffret-b replied:
Changed it.

Wadeck replied:
> many places causing huge issues

@jtnord Could you elaborate for future readers and myself?

jtnord replied:
if someone adapts a field with the same name then all classes using that field will be fudged at runtime (kind of OK).  but also any class that declares a field with the same name will be re-written.   The more classes that get rewritten the more likely you are to hit a bug in the bytecode-compatability-transformer.  and when you hit a bug in there have fun!

there is no need for these fields to be public.  runtime users can use groovy to change it (which does not care), test code can set a system property before instantiating the class, or if it wants to change it at runtime use some reflection, or make the property protected or other parts.  but adding to a public API a property to control switching removes your ability to refactor later.

jeffret-b replied:
@jtnord, is your concern here only related to jenkins-test-harness? I think your concern here is more general but I'm not certain.

jtnord replied:
my concern is you don't need public variables  don't create them.

I have not seen a need for this to be a public variable.

Wadeck replied:
It seems there is a lot of such behavior in Jenkins core. The argument is totally valid.

jtnord commented on core/src/main/java/jenkins/security/ExceptionTranslationFilter.java:
```diff
@@ -122,9 +126,9 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
-			if (ex.getRootCause() instanceof AuthenticationException
-					|| ex.getRootCause() instanceof AccessDeniedException) {
-				handleException(request, response, chain, (AcegiSecurityException) ex.getRootCause());
+			AcegiSecurityException securityException = containsAccessException(ex);
```
this is a larger change.  if I handle the access denied exception but wrap it with another exception that then causes an exception the current logic is to actually show the top level exception.  wheras now it punts to the access denied exception.  whats the case for looking deeper than one level?

jeffret-b replied:
The `/configureSecurity` case @Wadeck described earlier is one example, as demonstrated in this new test [here](https://github.com/jenkinsci/jenkins/blob/232edf5670ef254376444777fe016f80f8e9f9c4/test/src/test/java/jenkins/security/StackTraceSuppressionTest.java#L74). The `AccessDeniedException2` is wrapped several levels deep in `RuntimeException`. In contrast, `/manage` comes through without the deeper check as shown in this new test [here](https://github.com/jenkinsci/jenkins/blob/232edf5670ef254376444777fe016f80f8e9f9c4/test/src/test/java/jenkins/security/StackTraceSuppressionTest.java#L58). The choice here is basically to let it be caught by the unhandled exception code or the access denied exception code. It's more helpful when it's the latter.

Wadeck replied:
I imagine we can thank jelly for the wrapping there :p

jtnord replied:
my gut tells me this is wrong, jelly just adds the servlet exception so one level which is what this code originally did.
if you are absolutely sure otherwise  then I believe the previous code loop should be removed and collapsed with this. (pretty sure both of the previous are subtypes of this exception off the top of my head)

jeffret-b replied:
Caught at line 128 of ExceptionTranslationFilter when a non-administrator user goes to `/configureSecurity` by `getCause()`:
ServletException
  RuntimeException
    JellyTagException
      AccessDeniedException2

At the same point, `ex.getRootCause()` gives a RuntimeException.

Wadeck replied:
@jtnord is the answer satisfying you or do you need something else? :)

jtnord replied:
I would argue that is a bug in the code - the jelly should be hidden behind a permission check. 
`<layout permission=Administer>...`

jtnord replied:
looks like there is a permission check - so is this a bug against groovy views?

jeffret-b replied:
I backed out the bit to check embedded exceptions.

jeffret-b committed:
Use the stapler incremental now that's republished.

jeffret-b committed:
Use the stapler incremental now that's republished.

jeffret-b committed:
Don't use public static boolean property.

jeffret-b committed:
Don't use public static boolean property.

jtnord commented on test/src/test/java/jenkins/security/StackTraceSuppressionTest.java:
```diff
@@ -50,8 +50,8 @@
     public void setup() {
-        Jenkins.SHOW_STACK_TRACE = false;
-        HttpResponses.SHOW_STACK_TRACE = false;
+        System.setProperty("jenkins.model.Jenkins.SHOW_STACK_TRACE", "false");
```
should also be cleared in a tear down `@After` (possibly static) - incase the last test to run sets it to true and we reuse surefire for other tests (which is the default and I did not see changed anywhere in the pom - but also when running multiple tests from the IDE).  

```suggestion
        System.clearProperty("jenkins.model.Jenkins.SHOW_STACK_TRACE");
```

jeffret-b committed:
Adjust tests.

jeffret-b committed:
Adjust tests.

jeffret-b committed:
Don't handle embedded security exceptions.

jeffret-b committed:
Don't handle embedded security exceptions.

jtnord commented on core/src/main/java/jenkins/security/ExceptionTranslationFilter.java:
```diff
@@ -235,4 +238,20 @@ public void init(FilterConfig filterConfig) throws ServletException {
-}
\ No newline at end of file
+
+	private AcegiSecurityException containsAccessException(Exception exception) {
```
unused now?

jeffret-b replied:
Good catch. Removed. Thanks for the review!

jtnord reviewed (approved):
lgtm, looks like an unused private method remains

jeffret-b committed:
Remove now unused method.

jeffret-b committed:
Remove now unused method.

jeffret-b committed:
Adopt Daniel's suggestion for tying Stapler change to Jenkins.

And remove duplication in tests.

jeffret-b committed:
Adopt Daniel's suggestion for tying Stapler change to Jenkins.

And remove duplication in tests.

jeffret-b commented:
I updated to use a different mechanism for the associated stapler change in stapler/stapler#176. Daniel recommended allowing the app (Jenkins) to control the HttpResponses error response, showing the "oops" page. I re-worked the stapler PR and then adapted this one to that. While doing so, I eliminated some duplication in the new tests.

If anyone wants to re-review for the changes to support the stapler change, please do so. Otherwise, I think this has sufficient reviews.

I'll mark this as on-hold while we get the stapler dependency finalized.

Wadeck commented on core/src/main/java/hudson/init/impl/InstallUncaughtExceptionHandler.java:
```diff
@@ -54,6 +50,42 @@ public static void init(final Jenkins j) throws IOException {
+    public static class ErrorCustomizer implements HttpResponses.ErrorCustomizer {
+        @CheckForNull
+        @Override
+        public HttpResponses.HttpResponseException handleError(int i, Throwable throwable) {
```
```suggestion
        public HttpResponses.HttpResponseException handleError(int code, Throwable cause) {
```

Wadeck reviewed (approved):
Could you provide the current status of the PR concerning the "four different error handling pages" and where they are caught?

The recent change with the customizer improves the coverage of types and that's nice for the user experience (to not have to know there is a hidden Stapler)

Not tested manually with the changes.
⚠️ We need to be careful with changes that are touching the exception handling, to be sure they are working by testing manually. It's so easy to jump into situation where the handling is throwing => infinite loop

daniel-beck commented:
> Could you provide the current status of the PR concerning the "four different error handling pages" and where they are caught?

Nothing changed since https://github.com/jenkinsci/jenkins/pull/4389#issuecomment-575734123 except that case 3 looks like case 4 now.

jeffret-b committed:
Adjust naming.

jeffret-b committed:
Adjust naming.

jeffret-b commented:
Different conditions:

1. Access Denied. This looks good and doesn't need any changes. One way to encounter this error is to log in as a user without Administer permission and try to access http://[JENKINS_HOST]/manage. See a proposed test here. I'm not aware of any way to enable full stack trace.
2. Access Denied (embedded exception). In some cases the access denied exception is embedded some layers deep. One example here is to access the http://[JENKINS_HOST]/configureSecurity page. In a discussion here this was considered a defect. This is outside the scope of this PR. Currently this is handled as an uncaught exception. (See item 5.)
 3. Nonexistent page. Wadeck gives an example above, http://[jenkins-url]/404 . This is handled at the Stapler level and presents a simple, textual page. Changing this is outside the scope of this PR. Full stack trace is enabled by setting Dispatcher.TRACE = true, which may be the default in some dev / testing environments.
4. Nonexistent adjunct. This is handled via changes here and in the companion at stapler/stapler#176. This PR contains a test for this case, which is justified because part of the handling occurs in core. This redirects to the "oops" page. By default the stack trace is not shown. If the property "jenkins.model.Jenkins.SHOW_STACK_TRACE" is set to true, the stack trace is shown.
5. Uncaught error in Jenkins. This is the page connected with this PR. This redirects to the "oops" page. By default the stack trace is not shown. If the property "jenkins.model.Jenkins.SHOW_STACK_TRACE" is set to true, the stack trace is shown.

jeffret-b committed:
Use released Stapler.

jeffret-b committed:
Use released Stapler.

jeffret-b commented:
I updated to use the released version of Stapler. This should now be ready to go so I'm removing the on-hold label.

jeffret-b commented:
Hm... test failures. Don't know if they're related. I'll put it on hold until I can resolve them.

jeffret-b committed:
Pass the HTTP status code on to the oops page.

If we didn't get a code then continue to use 500. If we did, then use it.

jeffret-b committed:
Merge branch 'suppressStack' of github.com:jeffret-b/jenkins into suppressStack

jeffret-b commented:
Updated to pass the HTTP status code through the "oops" page when it is available.

Still waiting on a release / incrementals in jenkins-test-harness to deal with additional test issues.

jeffret-b committed:
Update to use newer jenkins-test-harness.

Use one that enables stack traces for tests that rely on it for easier development and troubleshooting.
Only in incrementals so far.

jeffret-b commented:
Don't know what happened to the build. Try to kick off another one.

jeffret-b committed:
Show stack trace for jetty:run.

jeffret-b commented:
Since the approach has stabilized and I'm already showing stack traces for tests, I went ahead and added @timja's suggestion to show the stack trace with `jetty:run`.

timja commented:
@jeffret-b can this be un on-holded now? I believe it was put on hold for test failures which are since fixed. 

Looks like the last 2 builds hit flakey tests, but it's now green.

daniel-beck commented:
When an "Apply" form submission results in an error, that error used to be shown in a "popup" style modal.

The current implementation breaks the "UI cleanup" happening there, so that no useful information is shown anymore. Additionally, the modal was always large, but it's clearly too large now, given what little content is left.

### Steps to reproduce

1. Create a freestyle job
2. Enter invalid nonsense in the "Build periodically" schedule
3. Press "Apply"

### Expected result

A fairly small modal that shows the generic error message and the error ID

### Actual result

A giant modal that doesn't show the error ID:

> ![Screenshot](https://user-images.githubusercontent.com/1831569/74596517-a55a8480-5050-11ea-9fe6-ec2ab39b366e.png)

daniel-beck reviewed (changes requested):
https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/lib/form/apply/apply.js should be updated correspondingly.

daniel-beck commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -5132,6 +5133,12 @@ private static void computeVersion(ServletContext context) {
+    public boolean shouldShowStackTrace() {
+        // Used by oops.jelly
+        return Boolean.getBoolean(Jenkins.class.getName() + ".SHOW_STACK_TRACE");
+    }
```
Given https://github.com/jenkinsci/jenkins-test-harness/pull/201/files#diff-1d6816bd6f9430280c5d8b7853286b1dR2688, I wonder whether a more classic approach (non-final field that is initialized from the system property) isn't a better approach to this.

`Jenkins.SHOW_ERROR_STACK_TRACES = true;` would be nicer over there.

daniel-beck replied:
Also https://github.com/jenkinsci/jenkins/pull/4389#pullrequestreview-359391807

jeffret-b replied:
See this comment: https://github.com/jenkinsci/jenkins/pull/4389#discussion_r369716605 . One approach would be to make it private, but that makes testing more difficult. On a different discussion Stephen Connolly presented an elaborate way to deal with that problem.

It's going to be difficult to have jenkins-test-harness do `Jenkins.SHOW_ERROR_STACK_TRACES = true;` because then JTH will depend upon a new Jenkins version and the new Jenkins needs the new JTH released for the tests to pass.

Wadeck replied:
Checking existence of the field in JTH using reflection will cover both the previous Jenkins version without that field and the new one with it, being private.
(IIUC Java 11 supports this as long as it's not a final keyword removal :trollface:)

> On a different discussion Stephen Connolly presented an elaborate way to deal with that problem.

Could you provide a link or at least a summary?

daniel-beck reviewed (changes requested):
Requesting another change. Reading through a neighboring PR, it reminded me that we use `SystemProperties` wrapper to access such values, as it also allows other sources to these options besides just actual system properties.

Then, this applies (from @oleg-nenashev  in https://github.com/jenkinsci/jenkins/pull/4463#pullrequestreview-357490289):

> Also keep in mind that `SystemProperties` API is rather designed for system properties which values are calculated once on restart. It involves calls like `ServletContext` checks, and I would not guarantee that it returns results immediately in all cases. Be careful when using the live switch in the frontend. If it is really needed, maybe using a `private static boolean` cache is preferable

So a field initialized to a `SystemProperty#getBoolean` seems like the way to go here.

jeffret-b commented:
@timja, it's been on hold waiting for the companion PR in jenkins-test-harness. With Daniel's latest comments, I'll put that on hold also, until I've addressed his concerns.

jeffret-b committed:
Adjust jelly for popup.

jeffret-b commented:
That popup is another part of Jenkins that I've never seen. It's a weird popup. It doesn't support many features. I've adjusted it so that it shows the logging ID and centers the line. I haven't been able to figure out how to adjust the size yet. If anyone has any pointers on how or where that is managed, I'd appreciate it.

daniel-beck commented:
> I haven't been able to figure out how to adjust the size yet. If anyone has any pointers on how or where that is managed, I'd appreciate it.

https://github.com/jenkinsci/jenkins/blob/fdfcbc3c7f55746878000f2e4b5599abada7ea73/core/src/main/resources/lib/form/apply/apply.js#L57-L69

jeffret-b committed:
Change popup dialog size.

jeffret-b commented:
<details>
  <summary>Adjusted the popup size</summary>
  
![Uploading Screen Shot 2020-02-21 at 11.19.09 AM.png…]()

</details>

jeffret-b commented:
<details>
  <summary>Adjusted the popup size</summary>
  
<img width="1198" alt="Screen Shot 2020-02-21 at 11 19 09 AM" src="https://user-images.githubusercontent.com/37345299/75060659-7a32c380-549c-11ea-902d-67ff54acbced.png">

</details>

daniel-beck commented on core/src/main/resources/lib/form/apply/apply.js:
```diff
@@ -40,10 +40,15 @@ Behaviour.specify("INPUT.apply-button", 'apply', 0, function (e) {
                     var doc = target.contentDocument || target.contentWindow.document;
                     var error = doc.getElementById('error-description');
+                    var r = YAHOO.util.Dom.getClientRegion();
+                    var contentHeight = r.height*1/5;
```
```suggestion
                    var contentHeight = r.height/5;
```

daniel-beck commented on core/src/main/resources/lib/form/apply/apply.js:
```diff
@@ -40,10 +40,15 @@ Behaviour.specify("INPUT.apply-button", 'apply', 0, function (e) {
                     var error = doc.getElementById('error-description');
+                    var r = YAHOO.util.Dom.getClientRegion();
+                    var contentHeight = r.height*1/5;
+                    var contentWidth = r.width*1/2;
```
```suggestion
                    var contentWidth = r.width/2;
```

jeffret-b committed:
Adjust math operations for clarity.

Co-Authored-By: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

daniel-beck commented:
Is it deliberate that the 'apply' error dialog doesn't show the stack trace even when the system property is set to show stack traces?

Not requesting this change, just asking.

daniel-beck reviewed (approved):
Still not a fan of the messaging on the screen, but that can be massaged later. This accomplishes the goal without any brokenness I was able to discover 👍

jeffret-b commented:
No intentional, but that popup is pretty weird. It's limited in what it understands. It doesn't show the rage image or the "Oops!" message. I couldn't get it to show the stack trace, at least not with the effort I've put into it so far. I didn't figure it was worth the effort to try and show the trace.

If you click the "Save" button rather than the "Apply" you get the regular oops page with the image and possibly the stack trace.

jeffret-b commented:
This needs JTH update and dependency update and then it should be ready to go.

daniel-beck commented:
> that popup is pretty weird. It's limited in what it understands. It doesn't show the rage image or the "Oops!" message. I couldn't get it to show the stack trace, at least not with the effort I've put into it so far

All of that is done by https://github.com/jenkinsci/jenkins/blob/a9c391073a02315955fa6f368f368add54d264b6/core/src/main/resources/lib/form/apply/apply.js#L42 and https://github.com/jenkinsci/jenkins/blob/a9c391073a02315955fa6f368f368add54d264b6/core/src/main/resources/lib/form/apply/apply.js#L59

jeffret-b commented:
We could do more to make that look nicer and add the stack trace to it if we want. I doubt it's worth it.

daniel-beck commented:
> more to make that look nicer

Besides the sizing challenge (larger with stack trace), all that's needed AFAICT is a new container `div` on the error page, and referencing that from `apply.js`. Not needed, but not a big challenge either. Like my other feedback, can be deferred.

daniel-beck committed:
Merge branch 'master' into suppressStack

daniel-beck commented:
@timja I think this is on-hold because of the incrementals dependency.

jeffret-b committed:
Merge branch 'master' into suppressStack

jeffret-b committed:
Merge branch 'suppressStack' of github.com:jeffret-b/jenkins into suppressStack

jeffret-b committed:
Merge branch 'master' into suppressStack

jeffret-b committed:
Adjust test for new permission changes.

Manage and system read permissions.

jeffret-b committed:
Merge branch 'suppressStack' of github.com:jeffret-b/jenkins into suppressStack

jeffret-b commented:
This one is on-hold because of the incrementals dependency to jenkins-test-harness. If we can get that one released, I can update this one and then this is ready.

timja commented on test/pom.xml:
```diff
@@ -70,7 +70,7 @@ THE SOFTWARE.
       <groupId>${project.groupId}</groupId>
       <artifactId>jenkins-test-harness</artifactId>
-      <version>2.61</version>
+      <version>2.62-rc1311.681a1adc1757</version>
```
released
```suggestion
      <version>2.62</version>
```

timja committed:
Update test/pom.xml

jeffret-b commented:
Thanks for the help with the JTH release, @timja! Once this builds with the released version, we should be able to remove the on-hold label.

jeffret-b commented:
Looks this is all ready to go. Unless there are any objections we should go ahead and merge this.
